cmd/dependencyverifier/dependencyverifier.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+type Unwanted struct {
+	// things we want to stop referencing
+	Spec UnwantedSpec `json:"spec"`
+	// status of our unwanted dependencies
+	Status UnwantedStatus `json:"status"`
+}
+
+type UnwantedSpec struct {
+	// TODO implement checks for RootModules
+	// module names/patterns of root modules whose dependencies should be considered direct references
+	RootModules []string `json:"rootModules"`
+
+	// module names we don't want to depend on, mapped to an optional message about why
+	UnwantedModules map[string]string `json:"unwantedModules"`
+}
+
+type UnwantedStatus struct {
+	// TODO implement checks for Vendored
+	// unwanted modules we still vendor, based on vendor/modules.txt content.
+	// eliminating things from this list is good.
+	Vendored []string `json:"vendored"`
+
+	// TODO implement checks for add DirectReferences
+	// unwanted modules we still directly reference from spec.roots, based on `go mod graph` content.
+	// eliminating things from this list is good.
+	DirectReferences []string `json:"directReferences"`
+
+	// unwanted modules indirectly referenced from modules other than spec.roots, based on `go mod graph` content.
```

Comment:
Feedback from the go team and behavior change in go1.17 indicate `go mod graph` isn't the right tool to determine direct/indirect dependencies (https://github.com/kubernetes/kubernetes/issues/104461). However, since `go list` does not return results for all build tags, there's not a clear alternative at the moment.

Reply:
Ugh, so how can we make progress here? :(

Reply:
hoisting from slack:

```
we can use go mod graph to get the full set of dependencies
which is what the tool is using now
but to resolve the "TODOs" distinguishing direct/indirect, it won't help us
so it would still be useful for keeping an unwanted dependency we've eradicated from creeping back in
but it won't be useful for keeping one that is currently an indirect dep from creeping in as a direct dep (edited) 
```